src/hostname/hostnamed.c

```diff
@@ -408,6 +409,62 @@ static int property_get_chassis(
 
         return sd_bus_message_append(reply, "s", name);
 }
+static const char * const dmi_invalid_list[] = {
+        "To be filled by OEM",
+        "TBD by OEM",
+        "Default String",
+        "Board Manufacturer",
+        "Board Vendor Name",
+        "Board Product Name",
+        NULL,	/* terminator */
+};
+
+static int is_dmi_valid(const char *field)
+{
+        int i = 0;
+
+        while (dmi_invalid_list[i]) {
+                if (strstr(field, dmi_invalid_list[i]))
+                        return 0;
+                i++;
+        }
+
+        return 1;
+}
+
+static int property_get_model(
+                sd_bus *bus,
+                const char *path,
+                const char *interface,
+                const char *property,
+                sd_bus_message *reply,
+                void *userdata,
+                sd_bus_error *error) {
+
+        char *is_lenovo, *model;
+        int r;
+
+        /* We usually want product_name, except for Lenovo. */
+        r = read_one_line_file("/sys/class/dmi/id/sys_vendor", &is_lenovo);
+        if (r < 0)
+                return -ENOENT;
+
+        if (strcasecmp(is_lenovo, "lenovo") == 0) {
+                free(is_lenovo);
+                r = read_one_line_file("/sys/class/dmi/id/product_version", &model);
+                if (r < 0)
+                        return -ENOENT;
+        } else {
+                r = read_one_line_file("/sys/class/dmi/id/product_name", &model);
+                if (r < 0)
+                        return -ENOENT;
+        }
```

Comment:
hmm, i really don't like the explicit check for lenovo here, this is not going to scale...

Maybe do it like this: if product_family exists and is valid use that, otherwise use product_name?

Reply:
Product family should be in platforms supporting SMBIOS 2.4 and later.  See table 10.
3.0 spec here: https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.0.0.pdf

Reply:
However I would like to mention that product family doesn't make sense on Dell platforms.  Defaulting to that will show less interesting stuff.
```
$ cat product_name product_family product_sku
XPS 13 7390
XPS
0962
```